enterprise_access/apps/content_assignments/api.py

```diff
@@ -21,28 +21,161 @@ def get_assignment_configuration(uuid):
+        'no-change': [LearnerContentAssignment records that matched
+                      the provided (config, learner, content) combination,
+                      but were already in an 'allocated' or 'accepted' state],
+      }
+      ```
+
+    """
+    # Fetch any existing assignments for all pairs of (learner, content) in this assignment config.
+    existing_assignments = get_assignments_by_learner_email_and_content(
+        assignment_configuration,
+        learner_emails,
+        content_key,
+    )
+
+    # Existing Assignments in consideration by state
+    already_allocated_or_accepted = []
+    cancelled_or_errored_to_update = []
+
+    # Maintain a set of emails with existing records - we know we don't have to create
+    # new assignments for these.
+    learner_emails_with_existing_assignments = set()
+
+    # Split up the existing assignment records by state
+    for assignment in existing_assignments:
+        learner_emails_with_existing_assignments.add(assignment.learner_email)
+        if assignment.state in LearnerContentAssignmentStateChoices.REALLOCATE_STATES:
+            assignment.content_quantity = content_price_cents
+            assignment.state = LearnerContentAssignmentStateChoices.ALLOCATED
+            cancelled_or_errored_to_update.append(assignment)
+        else:
+            already_allocated_or_accepted.append(assignment)
+
+    # Bulk update and get a list of refreshed objects
+    updated_assignments = _update_and_refresh_assignments(
+        cancelled_or_errored_to_update,
+        ['content_quantity', 'state']
+    )
+
+    # Narrow down creation list of learner emails
+    learner_emails_for_assignment_creation = set(learner_emails) - learner_emails_with_existing_assignments
+
+    # Initialize and save LearnerContentAssignment instances for each of them
+    created_assignments = _create_new_assignments(
+        assignment_configuration,
+        learner_emails_for_assignment_creation,
+        content_key,
+        content_price_cents,
+    )
+
+    # Return a mapping of the action we took to lists of relevant assignment records.
+    return {
+        'updated': updated_assignments,
+        'created': created_assignments,
+        'no_change': already_allocated_or_accepted,
+    }
+
+
+def _update_and_refresh_assignments(assignment_records, fields_changed):
+    """
+    Helper to bulk save the given assignment_records
+    and refresh their state from the DB.
+    """
+    # Save the assignments to update
+    LearnerContentAssignment.bulk_update(assignment_records, fields_changed)
+
+    # Get a list of refreshed objects that we just updated
+    return LearnerContentAssignment.objects.filter(
+        uuid__in=[record.uuid for record in assignment_records],
+    )
+
+
+def _create_new_assignments(assignment_configuration, learner_emails, content_key, content_price_cents):
+    """
+    Helper to bulk save new LearnerContentAssignment instances.
+    """
```

Comment:
In the future, is this the function that also kicks off linking/notification tasks?

Reply:
Yeah, either here, or do it directly from `allocated_assignments()` above, calling something like `notify_learner()` on the `updated_assignments` and `created_assignments` records.